File: app/src/main/java/com/jtmcn/archwiki/viewer/MainActivity.java
package com.jtmcn.archwiki.viewer;

import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.ShareActionProvider;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.webkit.WebSettings;
import android.widget.ProgressBar;

import com.jtmcn.archwiki.viewer.data.SearchResult;
import com.jtmcn.archwiki.viewer.data.SearchResultsBuilder;
import com.jtmcn.archwiki.viewer.data.WikiPage;
import com.jtmcn.archwiki.viewer.tasks.Fetch;
import com.jtmcn.archwiki.viewer.tasks.FetchUrl;
import com.jtmcn.archwiki.viewer.utils.AndroidUtils;
import com.jtmcn.archwiki.viewer.utils.SettingsUtils;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MainActivity extends AppCompatActivity implements FetchUrl.OnFinish<List<SearchResult>> {
	public static final String TAG = MainActivity.class.getSimpleName();
	@BindView(R.id.wiki_view) WikiView wikiViewer;
	@BindView(R.id.toolbar) Toolbar toolbar;
	private ShareActionProvider shareActionProvider;
	private SearchView searchView;
	private MenuItem searchMenuItem;
	private List<SearchResult> currentSuggestions;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);
		ButterKnife.bind(this);

		setSupportActionBar(toolbar);

		ProgressBar progressBar = ButterKnife.findById(this, R.id.progress_bar);
		wikiViewer.buildView(progressBar, getSupportActionBar());

		handleIntent(getIntent());
	}

	@Override
	protected void onResume() {
		super.onResume();
		updateWebSettings();
	}

	@Override
	protected void onNewIntent(Intent intent) {
		handleIntent(intent);
	}

	private void handleIntent(Intent intent) {
		if (intent == null) {
			return;
		}

		if (Intent.ACTION_SEARCH.equals(intent.getAction())) {
			String query = intent.getStringExtra(SearchManager.QUERY);
			wikiViewer.passSearch(query);
			hideSearchView();
		} else if (Intent.ACTION_VIEW.equals(intent.getAction())) {
			final String url = intent.getDataString();
			wikiViewer.wikiClient.shouldOverrideUrlLoading(wikiViewer, url);
		}
	}

	/**
	 * Update the font size used in the webview.
	 */
	public void updateWebSettings() {
		WebSettings webSettings = wikiViewer.getSettings();
		int fontSize = SettingsUtils.getFontSize(this);

		//todo this setting should be changed to a slider, remove deprecated call
		// deprecated method must be used until Android API 14
		// https://developer.android.com/reference/android/webkit/WebSettings.TextSize.html#NORMAL
		switch (fontSize) {
			case 0:
				webSettings.setTextSize(WebSettings.TextSize.SMALLEST); //50%
				break;
			case 1:
				webSettings.setTextSize(WebSettings.TextSize.SMALLER); //75%
				break;
			case 2:
				webSettings.setTextSize(WebSettings.TextSize.NORMAL); //100%
				break;
			case 3:
				webSettings.setTextSize(WebSettings.TextSize.LARGER); //150%
				break;
			case 4:
				webSettings.setTextSize(WebSettings.TextSize.LARGEST); //200%
				break;
		}
	}

	@Override
	public boolean onPrepareOptionsMenu(Menu menu) {
		SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
		searchMenuItem = menu.findItem(R.id.menu_search);
		searchView = (SearchView) MenuItemCompat.getActionView(searchMenuItem);
		searchView.setOnQueryTextFocusChangeListener(new View.OnFocusChangeListener() {
			@Override
			public void onFocusChange(View v, boolean hasFocus) {
				if (!hasFocus) {
					hideSearchView();
				}
			}
		});
		searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
		searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
			@Override
			public boolean onQueryTextSubmit(String query) {
				wikiViewer.passSearch(query);
				return false;
			}

			@Override
			public boolean onQueryTextChange(String newText) {
				if (newText.isEmpty()) {
					setCursorAdapter(new ArrayList<SearchResult>());
					return true;
				} else {
					String searchUrl = SearchResultsBuilder.getSearchQuery(newText);
					Fetch.search(MainActivity.this, searchUrl);
					return true;
				}
			}
		});

		searchView.setOnSuggestionListener(new SearchView.OnSuggestionListener() {
			@Override
			public boolean onSuggestionSelect(int position) {
				return false;
			}

			@Override
			public boolean onSuggestionClick(int position) {
				SearchResult searchResult = currentSuggestions.get(position);
				Log.d(TAG, "Opening '" + searchResult.getPageName() + "' from search suggestion.");
				wikiViewer.wikiClient.shouldOverrideUrlLoading(wikiViewer, searchResult.getPageUrl());
				hideSearchView();
				return true;
			}
		});
		return true;
	}

	public void hideSearchView() {
		searchMenuItem.collapseActionView();
		wikiViewer.requestFocus(); //pass control back to the wikiview
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.menu, menu);
		MenuItem share = menu.findItem(R.id.menu_share);
		shareActionProvider = (ShareActionProvider) MenuItemCompat.getActionProvider(share);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case R.id.menu_share:
				WikiPage wikiPage = wikiViewer.getCurrentWebPage();
				if (wikiPage != null) {
					Intent intent = AndroidUtils.shareText(wikiPage.getPageTitle(), wikiPage.getPageUrl(), this);
					shareActionProvider.setShareIntent(intent);
				}
				break;
			case R.id.refresh:
				wikiViewer.onRefresh();
				break;
			case R.id.menu_settings:
				startActivity(new Intent(this, PreferencesActivity.class));
				break;
			case R.id.exit:
				finish();
				break;
		}
		return true;
	}


	@Override
	public void onFinish(List<SearchResult> results) {
		currentSuggestions = results;
		setCursorAdapter(currentSuggestions);
	}

	private void setCursorAdapter(List<SearchResult> currentSuggestions) {
		searchView.setSuggestionsAdapter(
				SearchResultsAdapter.getCursorAdapter(this, currentSuggestions)
		);
	}
}

File: app/src/main/java/com/jtmcn/archwiki/viewer/MainActivity.kt
package com.jtmcn.archwiki.viewer

import android.app.SearchManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import com.jtmcn.archwiki.viewer.data.SearchResult
import com.jtmcn.archwiki.viewer.data.getSearchQuery
import com.jtmcn.archwiki.viewer.tasks.Fetch
import com.jtmcn.archwiki.viewer.utils.getTextZoom
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.toolbar.*
import timber.log.Timber
import java.util.*

class MainActivity : AppCompatActivity() {
    private lateinit var searchView: SearchView
    private lateinit var searchMenuItem: MenuItem
    private var currentSuggestions: List<SearchResult>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        setSupportActionBar(toolbar)

        wikiViewer.buildView(progressBar, supportActionBar)

        handleIntent(intent)
    }

    override fun onResume() {
        super.onResume()
        updateWebSettings()
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        handleIntent(intent)
    }

    private fun handleIntent(intent: Intent?) {
        if (intent == null) {
            return
        }

        if (Intent.ACTION_SEARCH == intent.action) {
            val query = intent.getStringExtra(SearchManager.QUERY)
            wikiViewer.passSearch(query!!)
            hideSearchView()
        } else if (Intent.ACTION_VIEW == intent.action) {
            val url = intent.dataString
            wikiViewer.wikiClient.shouldOverrideUrlLoading(wikiViewer, url!!)
        }
    }

    /**
     * Update the font size used in the webview.
     */
    private fun updateWebSettings() {
        wikiViewer.settings.textZoom = getTextZoom(this)
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
        searchMenuItem = menu.findItem(R.id.menu_search)
        searchView = searchMenuItem.actionView as SearchView
        searchView.setOnQueryTextFocusChangeListener { _, hasFocus ->
            if (!hasFocus) {
                hideSearchView()
            }
        }
        searchView.setSearchableInfo(searchManager.getSearchableInfo(componentName))
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                wikiViewer.passSearch(query)
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                if (newText.isEmpty()) {
                    setCursorAdapter(ArrayList())
                    return true
                } else {
                    val searchUrl = getSearchQuery(newText)
                    Fetch.search({
                        currentSuggestions = it
                        setCursorAdapter(currentSuggestions)
                    }, searchUrl)
                    return true
                }
            }
        })

        searchView.setOnSuggestionListener(object : SearchView.OnSuggestionListener {
            override fun onSuggestionSelect(position: Int): Boolean {
                return false
            }

            override fun onSuggestionClick(position: Int): Boolean {
                val (pageName, pageURL) = currentSuggestions!![position]
                Timber.d("Opening '$pageName' from search suggestion.")
                wikiViewer.wikiClient.shouldOverrideUrlLoading(wikiViewer, pageURL)
                hideSearchView()
                return true
            }
        })
        return true
    }

    private fun hideSearchView() {
        searchMenuItem.collapseActionView()
        wikiViewer.requestFocus() //pass control back to the wikiview
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_share -> {
                val wikiPage = wikiViewer.currentWebPage
                if (wikiPage != null) {
                    val sharingIntent = Intent()
                    sharingIntent.type = TEXT_PLAIN_MIME
                    sharingIntent.action = Intent.ACTION_SEND
                    sharingIntent.putExtra(Intent.EXTRA_TITLE, wikiPage.pageTitle)
                    sharingIntent.putExtra(Intent.EXTRA_TEXT, wikiPage.pageUrl)
                    startActivity(Intent.createChooser(sharingIntent, null))
                }
            }
            R.id.refresh -> wikiViewer.onRefresh()
            R.id.menu_settings -> startActivity(Intent(this, PreferencesActivity::class.java))
            R.id.exit -> finish()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun setCursorAdapter(currentSuggestions: List<SearchResult>?) {
        searchView.suggestionsAdapter = SearchResultsAdapter.getCursorAdapter(this, currentSuggestions!!)
    }
}